{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2df488db-dcc1-4f70-b626-1609e7e6a1ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "import time\n",
    "from scipy.sparse import csr_matrix\n",
    "import anndata as an\n",
    "import scanpy as sc\n",
    "import pyranges as pr\n",
    "import psutil\n",
    "\n",
    "# import the script\n",
    "source_path = os.path.abspath(\"../scripts/\")\n",
    "sys.path.append(source_path)\n",
    "import make_anndata as mk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "b31d3e77-1fac-46b8-9b79-afa43520d3d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time taken to read the file: 184.47 seconds\n",
      "Memory usage: current 21.00 GB, difference +21.00 GB\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 2579 × 2756467\n",
       "    obs: 'bin_index', 'bin_start', 'bin_end', 'bin', 'chrom_bin', 'degree', 'genes', 'n_genes'\n",
       "    var: 'read_index', 'basename', 'mean_mapq', 'median_mapq', 'n_chromosomes', 'order', 'n_bins', 'read_length_bp', 'genes', 'n_genes'\n",
       "    uns: 'base_resolution', 'chrom_sizes', 'gdf', 'gene_map', 'intervals'\n",
       "    layers: 'H'"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fpath = \"/scratch/indikar_root/indikar1/shared_data/higher_order/anndata/population_mESC_1000000_raw.h5ad\"\n",
    "\n",
    "start_time = time.time()  # Record the start time\n",
    "adata = sc.read_h5ad(fpath)\n",
    "end_time = time.time()  # Record the end time\n",
    "print(f\"Time taken to read the file: {end_time - start_time:.2f} seconds\")\n",
    "sc.logging.print_memory_usage()\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1defdf4c-3f87-4d98-8f5f-c10f029d9d7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc3d1be3-b32e-4278-b324-9f03167787a6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bdddf88-2588-43fd-a5db-42e41d999c79",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f86f282-ecd3-4b89-a82c-20f1760d71ce",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cd926a7-f748-4a32-a9d3-ef7824328193",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a368287b-7bdf-4460-9532-52c8990d563d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "880bd167-9beb-47a6-a3a8-8fed84ecdeee",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b3440f7-881d-4bd8-85a6-3c3dbb76560e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a82e0531-8c51-4004-a59d-61736a668adc",
   "metadata": {},
   "outputs": [],
   "source": [
    "pore_c_path = \"/scratch/indikar_root/indikar1/shared_data/higher_order/pore_c/population_mESC.read_level.parquet\"\n",
    "resolution = 1000000\n",
    "chrom_path = \"/scratch/indikar_root/indikar1/shared_data/higher_order/reference/chrom_sizes.csv\"\n",
    "gene_path = \"/scratch/indikar_root/indikar1/shared_data/higher_order/reference/gene_table.parquet\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a9a804b4-fdd1-4946-ab20-fbf39c5d47c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------------------------------------------------------------\n",
      "------------------- Loading Pore-C Data --------------------\n",
      "------------------------------------------------------------\n",
      "Pore-C data shape:                  (14877807, 13)\n",
      "RAM usage at step 'Load Pore-C data': 2.41 GB\n"
     ]
    }
   ],
   "source": [
    "# Load the Pore-C data\n",
    "mk.print_section_header(\"Loading Pore-C Data\")\n",
    "df = pd.read_parquet(pore_c_path)\n",
    "df['value'] = 1\n",
    "mk.print_data_shape(\"Pore-C data\", df.shape)\n",
    "mk.print_memory_usage(\"Load Pore-C data\")\n",
    "\n",
    "n = df.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a24bb66e-4599-426c-b2ad-f42ac0874e3b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------------------------------------------------------------\n",
      "-------------- Creating Chromosome Intervals ---------------\n",
      "------------------------------------------------------------\n",
      "intervals.shape=(2642, 6)\n",
      "Chromosome intervals shape:         (2642, 6)\n"
     ]
    }
   ],
   "source": [
    "# Load the chromosome table\n",
    "mk.print_section_header(\"Creating Chromosome Intervals\")\n",
    "chrom, intervals = mk.create_chromosome_intervals(chrom_path, base_resolution=resolution)\n",
    "mk.print_data_shape(\"Chromosome intervals\", intervals.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8bcb7831-a145-48e6-9127-eff107c5661d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>bin</th>\n",
       "      <th>chrom_bin</th>\n",
       "      <th>bin_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1000000</td>\n",
       "      <td>196</td>\n",
       "      <td>0</td>\n",
       "      <td>chr2:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>2</td>\n",
       "      <td>1000000</td>\n",
       "      <td>2000000</td>\n",
       "      <td>197</td>\n",
       "      <td>1</td>\n",
       "      <td>chr2:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>2</td>\n",
       "      <td>2000000</td>\n",
       "      <td>3000000</td>\n",
       "      <td>198</td>\n",
       "      <td>2</td>\n",
       "      <td>chr2:2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>2</td>\n",
       "      <td>3000000</td>\n",
       "      <td>4000000</td>\n",
       "      <td>199</td>\n",
       "      <td>3</td>\n",
       "      <td>chr2:3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>2</td>\n",
       "      <td>4000000</td>\n",
       "      <td>5000000</td>\n",
       "      <td>200</td>\n",
       "      <td>4</td>\n",
       "      <td>chr2:4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom    start      end  bin  chrom_bin bin_name\n",
       "196     2        0  1000000  196          0   chr2:0\n",
       "197     2  1000000  2000000  197          1   chr2:1\n",
       "198     2  2000000  3000000  198          2   chr2:2\n",
       "199     2  3000000  4000000  199          3   chr2:3\n",
       "200     2  4000000  5000000  200          4   chr2:4"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intervals[intervals['chrom'] == '2'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6735f59a-6dac-4098-80ec-1a27b1dbf49e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df.shape=(14877807, 24)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_name</th>\n",
       "      <th>read_start</th>\n",
       "      <th>read_end</th>\n",
       "      <th>length_on_read</th>\n",
       "      <th>chrom</th>\n",
       "      <th>ref_start</th>\n",
       "      <th>ref_end</th>\n",
       "      <th>mapping_quality</th>\n",
       "      <th>basename</th>\n",
       "      <th>local_position</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_source</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>midpoint</th>\n",
       "      <th>is_tf</th>\n",
       "      <th>gene_overlap</th>\n",
       "      <th>is_pt_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8342354</th>\n",
       "      <td>3891ee6d-53d1-4ee0-ba2f-3d22291d4493</td>\n",
       "      <td>0</td>\n",
       "      <td>4916</td>\n",
       "      <td>4916</td>\n",
       "      <td>9</td>\n",
       "      <td>121048825</td>\n",
       "      <td>121057172</td>\n",
       "      <td>60</td>\n",
       "      <td>batch01</td>\n",
       "      <td>121052998</td>\n",
       "      <td>...</td>\n",
       "      <td>Ulk4</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>120784416</td>\n",
       "      <td>121106263</td>\n",
       "      <td>321847</td>\n",
       "      <td>120945339</td>\n",
       "      <td>False</td>\n",
       "      <td>8347</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14789315</th>\n",
       "      <td>66953ddf-e76d-4cdf-aaf8-be028a2d7b04</td>\n",
       "      <td>6352</td>\n",
       "      <td>9540</td>\n",
       "      <td>3188</td>\n",
       "      <td>19</td>\n",
       "      <td>26583825</td>\n",
       "      <td>26592064</td>\n",
       "      <td>60</td>\n",
       "      <td>batch01</td>\n",
       "      <td>26587944</td>\n",
       "      <td>...</td>\n",
       "      <td>Smarca2</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>26582449</td>\n",
       "      <td>26755722</td>\n",
       "      <td>173273</td>\n",
       "      <td>26669085</td>\n",
       "      <td>False</td>\n",
       "      <td>8239</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4110544</th>\n",
       "      <td>ad5b2240-893f-4ed0-a157-c2be66d8d754</td>\n",
       "      <td>0</td>\n",
       "      <td>1919</td>\n",
       "      <td>1919</td>\n",
       "      <td>4</td>\n",
       "      <td>127067225</td>\n",
       "      <td>127074760</td>\n",
       "      <td>60</td>\n",
       "      <td>batch04</td>\n",
       "      <td>127070992</td>\n",
       "      <td>...</td>\n",
       "      <td>Dlgap3</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>127062996</td>\n",
       "      <td>127130815</td>\n",
       "      <td>67819</td>\n",
       "      <td>127096905</td>\n",
       "      <td>False</td>\n",
       "      <td>7535</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10733063</th>\n",
       "      <td>3f354c45-5e48-4f6d-8c7e-05369432b344</td>\n",
       "      <td>762</td>\n",
       "      <td>5998</td>\n",
       "      <td>5236</td>\n",
       "      <td>12</td>\n",
       "      <td>8680599</td>\n",
       "      <td>8685932</td>\n",
       "      <td>60</td>\n",
       "      <td>batch04</td>\n",
       "      <td>8683265</td>\n",
       "      <td>...</td>\n",
       "      <td>Gm56531</td>\n",
       "      <td>havana_tagene</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>8649709</td>\n",
       "      <td>8691034</td>\n",
       "      <td>41325</td>\n",
       "      <td>8670371</td>\n",
       "      <td>False</td>\n",
       "      <td>5333</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13889148</th>\n",
       "      <td>d4626feb-16a2-4aac-8145-53e89b60bf7c</td>\n",
       "      <td>0</td>\n",
       "      <td>1821</td>\n",
       "      <td>1821</td>\n",
       "      <td>17</td>\n",
       "      <td>66243365</td>\n",
       "      <td>66248641</td>\n",
       "      <td>60</td>\n",
       "      <td>batch01</td>\n",
       "      <td>66246003</td>\n",
       "      <td>...</td>\n",
       "      <td>Twsg1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>66228966</td>\n",
       "      <td>66258221</td>\n",
       "      <td>29255</td>\n",
       "      <td>66243593</td>\n",
       "      <td>False</td>\n",
       "      <td>5276</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     read_name  read_start  read_end  \\\n",
       "8342354   3891ee6d-53d1-4ee0-ba2f-3d22291d4493           0      4916   \n",
       "14789315  66953ddf-e76d-4cdf-aaf8-be028a2d7b04        6352      9540   \n",
       "4110544   ad5b2240-893f-4ed0-a157-c2be66d8d754           0      1919   \n",
       "10733063  3f354c45-5e48-4f6d-8c7e-05369432b344         762      5998   \n",
       "13889148  d4626feb-16a2-4aac-8145-53e89b60bf7c           0      1821   \n",
       "\n",
       "          length_on_read chrom  ref_start    ref_end  mapping_quality  \\\n",
       "8342354             4916     9  121048825  121057172               60   \n",
       "14789315            3188    19   26583825   26592064               60   \n",
       "4110544             1919     4  127067225  127074760               60   \n",
       "10733063            5236    12    8680599    8685932               60   \n",
       "13889148            1821    17   66243365   66248641               60   \n",
       "\n",
       "         basename  local_position  ...  gene_name     gene_source  \\\n",
       "8342354   batch01       121052998  ...       Ulk4  ensembl_havana   \n",
       "14789315  batch01        26587944  ...    Smarca2  ensembl_havana   \n",
       "4110544   batch04       127070992  ...     Dlgap3  ensembl_havana   \n",
       "10733063  batch04         8683265  ...    Gm56531   havana_tagene   \n",
       "13889148  batch01        66246003  ...      Twsg1  ensembl_havana   \n",
       "\n",
       "            gene_biotype gene_start   gene_end gene_length   midpoint  is_tf  \\\n",
       "8342354   protein_coding  120784416  121106263      321847  120945339  False   \n",
       "14789315  protein_coding   26582449   26755722      173273   26669085  False   \n",
       "4110544   protein_coding  127062996  127130815       67819  127096905  False   \n",
       "10733063          lncRNA    8649709    8691034       41325    8670371  False   \n",
       "13889148  protein_coding   66228966   66258221       29255   66243593  False   \n",
       "\n",
       "          gene_overlap  is_pt_gene  \n",
       "8342354           8347        True  \n",
       "14789315          8239        True  \n",
       "4110544           7535        True  \n",
       "10733063          5333       False  \n",
       "13889148          5276        True  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = mk.merge_genes(df, gene_path)\n",
    "print(f\"{df.shape=}\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "62473f0e-8184-46de-bf7a-6c397f74e1e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n - df.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "930c92b6-2c1d-4935-9d89-913cc713a73c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------------------------------------------------------------\n",
      "------------ Joining Intervals with Pore-C Data ------------\n",
      "------------------------------------------------------------\n",
      "df.shape=(14877807, 32)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_name</th>\n",
       "      <th>read_start</th>\n",
       "      <th>read_end</th>\n",
       "      <th>length_on_read</th>\n",
       "      <th>chrom</th>\n",
       "      <th>ref_start</th>\n",
       "      <th>ref_end</th>\n",
       "      <th>mapping_quality</th>\n",
       "      <th>basename</th>\n",
       "      <th>local_position</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_overlap</th>\n",
       "      <th>is_pt_gene</th>\n",
       "      <th>bin_start</th>\n",
       "      <th>bin_end</th>\n",
       "      <th>bin</th>\n",
       "      <th>chrom_bin</th>\n",
       "      <th>bin_name</th>\n",
       "      <th>bin_overlap</th>\n",
       "      <th>read_index</th>\n",
       "      <th>bin_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7796001</th>\n",
       "      <td>3891ee6d-53d1-4ee0-ba2f-3d22291d4493</td>\n",
       "      <td>0</td>\n",
       "      <td>4916</td>\n",
       "      <td>4916</td>\n",
       "      <td>9</td>\n",
       "      <td>121048825</td>\n",
       "      <td>121057172</td>\n",
       "      <td>60</td>\n",
       "      <td>batch01</td>\n",
       "      <td>121052998</td>\n",
       "      <td>...</td>\n",
       "      <td>8347</td>\n",
       "      <td>True</td>\n",
       "      <td>121000000</td>\n",
       "      <td>122000000</td>\n",
       "      <td>1394</td>\n",
       "      <td>121</td>\n",
       "      <td>chr9:121</td>\n",
       "      <td>8347</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14148112</th>\n",
       "      <td>66953ddf-e76d-4cdf-aaf8-be028a2d7b04</td>\n",
       "      <td>6352</td>\n",
       "      <td>9540</td>\n",
       "      <td>3188</td>\n",
       "      <td>19</td>\n",
       "      <td>26583825</td>\n",
       "      <td>26592064</td>\n",
       "      <td>60</td>\n",
       "      <td>batch01</td>\n",
       "      <td>26587944</td>\n",
       "      <td>...</td>\n",
       "      <td>8239</td>\n",
       "      <td>True</td>\n",
       "      <td>26000000</td>\n",
       "      <td>27000000</td>\n",
       "      <td>2436</td>\n",
       "      <td>26</td>\n",
       "      <td>chr19:26</td>\n",
       "      <td>8239</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3788352</th>\n",
       "      <td>ad5b2240-893f-4ed0-a157-c2be66d8d754</td>\n",
       "      <td>0</td>\n",
       "      <td>1919</td>\n",
       "      <td>1919</td>\n",
       "      <td>4</td>\n",
       "      <td>127067225</td>\n",
       "      <td>127074760</td>\n",
       "      <td>60</td>\n",
       "      <td>batch04</td>\n",
       "      <td>127070992</td>\n",
       "      <td>...</td>\n",
       "      <td>7535</td>\n",
       "      <td>True</td>\n",
       "      <td>127000000</td>\n",
       "      <td>128000000</td>\n",
       "      <td>665</td>\n",
       "      <td>127</td>\n",
       "      <td>chr4:127</td>\n",
       "      <td>7535</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7462945</th>\n",
       "      <td>1cbc7b8c-2ad8-4365-a57c-037fc41af923</td>\n",
       "      <td>1305</td>\n",
       "      <td>6701</td>\n",
       "      <td>5396</td>\n",
       "      <td>8</td>\n",
       "      <td>21218739</td>\n",
       "      <td>21225694</td>\n",
       "      <td>1</td>\n",
       "      <td>batch04</td>\n",
       "      <td>21222216</td>\n",
       "      <td>...</td>\n",
       "      <td>-21218740</td>\n",
       "      <td>False</td>\n",
       "      <td>21000000</td>\n",
       "      <td>22000000</td>\n",
       "      <td>1163</td>\n",
       "      <td>21</td>\n",
       "      <td>chr8:21</td>\n",
       "      <td>6955</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7463448</th>\n",
       "      <td>9af4c1b9-a795-43a1-9669-55aff68e3b9d</td>\n",
       "      <td>426</td>\n",
       "      <td>5786</td>\n",
       "      <td>5360</td>\n",
       "      <td>8</td>\n",
       "      <td>21241017</td>\n",
       "      <td>21247152</td>\n",
       "      <td>60</td>\n",
       "      <td>batch03</td>\n",
       "      <td>21244084</td>\n",
       "      <td>...</td>\n",
       "      <td>-21241018</td>\n",
       "      <td>False</td>\n",
       "      <td>21000000</td>\n",
       "      <td>22000000</td>\n",
       "      <td>1163</td>\n",
       "      <td>21</td>\n",
       "      <td>chr8:21</td>\n",
       "      <td>6135</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     read_name  read_start  read_end  \\\n",
       "7796001   3891ee6d-53d1-4ee0-ba2f-3d22291d4493           0      4916   \n",
       "14148112  66953ddf-e76d-4cdf-aaf8-be028a2d7b04        6352      9540   \n",
       "3788352   ad5b2240-893f-4ed0-a157-c2be66d8d754           0      1919   \n",
       "7462945   1cbc7b8c-2ad8-4365-a57c-037fc41af923        1305      6701   \n",
       "7463448   9af4c1b9-a795-43a1-9669-55aff68e3b9d         426      5786   \n",
       "\n",
       "          length_on_read chrom  ref_start    ref_end  mapping_quality  \\\n",
       "7796001             4916     9  121048825  121057172               60   \n",
       "14148112            3188    19   26583825   26592064               60   \n",
       "3788352             1919     4  127067225  127074760               60   \n",
       "7462945             5396     8   21218739   21225694                1   \n",
       "7463448             5360     8   21241017   21247152               60   \n",
       "\n",
       "         basename  local_position  ...  gene_overlap  is_pt_gene  bin_start  \\\n",
       "7796001   batch01       121052998  ...          8347        True  121000000   \n",
       "14148112  batch01        26587944  ...          8239        True   26000000   \n",
       "3788352   batch04       127070992  ...          7535        True  127000000   \n",
       "7462945   batch04        21222216  ...     -21218740       False   21000000   \n",
       "7463448   batch03        21244084  ...     -21241018       False   21000000   \n",
       "\n",
       "            bin_end   bin chrom_bin  bin_name  bin_overlap  read_index  \\\n",
       "7796001   122000000  1394       121  chr9:121         8347           0   \n",
       "14148112   27000000  2436        26  chr19:26         8239           1   \n",
       "3788352   128000000   665       127  chr4:127         7535           2   \n",
       "7462945    22000000  1163        21   chr8:21         6955           3   \n",
       "7463448    22000000  1163        21   chr8:21         6135           4   \n",
       "\n",
       "          bin_index  \n",
       "7796001           0  \n",
       "14148112          1  \n",
       "3788352           2  \n",
       "7462945           3  \n",
       "7463448           3  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add the interval information\n",
    "mk.print_section_header(\"Joining Intervals with Pore-C Data\")\n",
    "df = mk.join_intervals_pyranges(df, intervals)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8ec27782-f242-4735-b7a7-8550522c1225",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n - df.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "41a92c18-817f-408a-9d4d-e20c549f8e0b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>basename</th>\n",
       "      <th>read_name</th>\n",
       "      <th>mean_mapq</th>\n",
       "      <th>median_mapq</th>\n",
       "      <th>n_chromosomes</th>\n",
       "      <th>order</th>\n",
       "      <th>n_bins</th>\n",
       "      <th>read_length_bp</th>\n",
       "      <th>genes</th>\n",
       "      <th>n_genes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>read_index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2585712</th>\n",
       "      <td>batch01</td>\n",
       "      <td>00001034-f0ab-41c2-885a-da17c2b836ae</td>\n",
       "      <td>34.000000</td>\n",
       "      <td>24.0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>378</td>\n",
       "      <td>Grm1;Lhfpl4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2099807</th>\n",
       "      <td>batch01</td>\n",
       "      <td>00001afb-6b62-4d01-be02-9509485330ab</td>\n",
       "      <td>59.000000</td>\n",
       "      <td>59.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>452</td>\n",
       "      <td>Adam12</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1156249</th>\n",
       "      <td>batch01</td>\n",
       "      <td>00002230-8080-4260-9522-dc39258a0698</td>\n",
       "      <td>60.000000</td>\n",
       "      <td>60.0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1695</td>\n",
       "      <td>Fggy;Gm830</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146652</th>\n",
       "      <td>batch01</td>\n",
       "      <td>00002fb3-034c-4900-aaa9-ca1a08f4fb6b</td>\n",
       "      <td>52.333333</td>\n",
       "      <td>60.0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1892</td>\n",
       "      <td>Csnka2ip;Gnal</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>970720</th>\n",
       "      <td>batch01</td>\n",
       "      <td>00003344-980a-405c-b539-eb0341ada8d2</td>\n",
       "      <td>35.000000</td>\n",
       "      <td>35.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>657</td>\n",
       "      <td>Upp2;Scfd2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           basename                             read_name  mean_mapq  \\\n",
       "read_index                                                             \n",
       "2585712     batch01  00001034-f0ab-41c2-885a-da17c2b836ae  34.000000   \n",
       "2099807     batch01  00001afb-6b62-4d01-be02-9509485330ab  59.000000   \n",
       "1156249     batch01  00002230-8080-4260-9522-dc39258a0698  60.000000   \n",
       "146652      batch01  00002fb3-034c-4900-aaa9-ca1a08f4fb6b  52.333333   \n",
       "970720      batch01  00003344-980a-405c-b539-eb0341ada8d2  35.000000   \n",
       "\n",
       "            median_mapq  n_chromosomes  order  n_bins  read_length_bp  \\\n",
       "read_index                                                              \n",
       "2585712            24.0              2      3       2             378   \n",
       "2099807            59.0              2      2       2             452   \n",
       "1156249            60.0              1      3       1            1695   \n",
       "146652             60.0              2      3       2            1892   \n",
       "970720             35.0              2      2       2             657   \n",
       "\n",
       "                    genes  n_genes  \n",
       "read_index                          \n",
       "2585712       Grm1;Lhfpl4        2  \n",
       "2099807            Adam12        1  \n",
       "1156249        Fggy;Gm830        2  \n",
       "146652      Csnka2ip;Gnal        2  \n",
       "970720         Upp2;Scfd2        2  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def create_var_df(df, var_names):\n",
    "  \"\"\"Creates a variable DataFrame from a DataFrame with read information.\n",
    "\n",
    "  Args:\n",
    "    df (pandas.DataFrame): DataFrame with read data including 'read_name', 'read_index', \n",
    "                           'mapping_quality', 'chrom', 'order', 'bin', and 'length_on_read'.\n",
    "    var_names (pandas.Index): Index of unique read names.\n",
    "\n",
    "  Returns:\n",
    "    pandas.DataFrame: DataFrame containing variable information (read-level summaries).\n",
    "  \"\"\"\n",
    "  var = df.copy()\n",
    "\n",
    "  gene_list = lambda x: \";\".join([i for i in set(x) if i != '-1'])\n",
    "  n_genes = lambda x: len([i for i in set(x) if i != '-1'])\n",
    "\n",
    "  var = var.groupby(['basename', 'read_name', 'read_index']).agg(\n",
    "      mean_mapq=('mapping_quality', 'mean'),\n",
    "      median_mapq=('mapping_quality', 'median'),\n",
    "      n_chromosomes=('chrom', 'nunique'),\n",
    "      order=('order', 'first'),\n",
    "      n_bins=('bin', 'nunique'),\n",
    "      read_length_bp=('length_on_read', 'sum'),\n",
    "      genes=('gene_name', gene_list),\n",
    "      n_genes=('gene_name', n_genes),\n",
    "  ).reset_index()\n",
    "\n",
    "  # Ensure proper sorting using var_names\n",
    "  var = var.set_index('read_index')\n",
    "  # var = var.reindex(var_names)\n",
    "  # var = var.reset_index()\n",
    "  # var = var.set_index('read_name')\n",
    "\n",
    "  return var\n",
    "\n",
    "\n",
    "var = create_var_df(df, \"\")\n",
    "var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b5d9c4ad-3030-48e6-bf24-f48020cb1d77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_name</th>\n",
       "      <th>bin_start</th>\n",
       "      <th>bin_end</th>\n",
       "      <th>bin</th>\n",
       "      <th>bin_index</th>\n",
       "      <th>chrom_bin</th>\n",
       "      <th>degree</th>\n",
       "      <th>genes</th>\n",
       "      <th>n_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10:10</td>\n",
       "      <td>10000000</td>\n",
       "      <td>11000000</td>\n",
       "      <td>1408</td>\n",
       "      <td>881</td>\n",
       "      <td>10</td>\n",
       "      <td>3356</td>\n",
       "      <td>Adgb;Gm48324;Grm1;4930567K20Rik;Gm48406;Gm4827...</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10:100</td>\n",
       "      <td>100000000</td>\n",
       "      <td>101000000</td>\n",
       "      <td>1498</td>\n",
       "      <td>1431</td>\n",
       "      <td>100</td>\n",
       "      <td>3109</td>\n",
       "      <td>Gm47956;Tmtc3;Gm47631;Gm4301;1700017N19Rik;Gm8...</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10:101</td>\n",
       "      <td>101000000</td>\n",
       "      <td>102000000</td>\n",
       "      <td>1499</td>\n",
       "      <td>1505</td>\n",
       "      <td>101</td>\n",
       "      <td>2865</td>\n",
       "      <td>Mgat4c;Gm19233;Gm26278</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10:102</td>\n",
       "      <td>102000000</td>\n",
       "      <td>103000000</td>\n",
       "      <td>1500</td>\n",
       "      <td>506</td>\n",
       "      <td>102</td>\n",
       "      <td>3055</td>\n",
       "      <td>Gm47033;Mgat4c;Rassf9;Gm26988;Gm5175;Gm47101;G...</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10:103</td>\n",
       "      <td>103000000</td>\n",
       "      <td>104000000</td>\n",
       "      <td>1501</td>\n",
       "      <td>446</td>\n",
       "      <td>103</td>\n",
       "      <td>2988</td>\n",
       "      <td>Gm26923;Gm47228;Gm47221;Gm47224;Gm21293;Gm4722...</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    bin_name  bin_start    bin_end   bin  bin_index  chrom_bin  degree  \\\n",
       "0   chr10:10   10000000   11000000  1408        881         10    3356   \n",
       "1  chr10:100  100000000  101000000  1498       1431        100    3109   \n",
       "2  chr10:101  101000000  102000000  1499       1505        101    2865   \n",
       "3  chr10:102  102000000  103000000  1500        506        102    3055   \n",
       "4  chr10:103  103000000  104000000  1501        446        103    2988   \n",
       "\n",
       "                                               genes  n_genes  \n",
       "0  Adgb;Gm48324;Grm1;4930567K20Rik;Gm48406;Gm4827...       13  \n",
       "1  Gm47956;Tmtc3;Gm47631;Gm4301;1700017N19Rik;Gm8...       23  \n",
       "2                             Mgat4c;Gm19233;Gm26278        3  \n",
       "3  Gm47033;Mgat4c;Rassf9;Gm26988;Gm5175;Gm47101;G...       11  \n",
       "4  Gm26923;Gm47228;Gm47221;Gm47224;Gm21293;Gm4722...       14  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def create_obs_df(df, obs_names):\n",
    "  \"\"\"Creates an observation DataFrame from a DataFrame and a sparse matrix.\n",
    "\n",
    "  Args:\n",
    "    df (pandas.DataFrame): DataFrame with genomic data and bin information.\n",
    "    X (scipy.sparse.csr_matrix): Sparse matrix representation of the data.\n",
    "\n",
    "  Returns:\n",
    "    pandas.DataFrame: DataFrame containing observation information.\n",
    "  \"\"\"\n",
    "  gene_list = lambda x: \";\".join([i for i in set(x) if i != '-1'])\n",
    "  n_genes = lambda x: len([i for i in set(x) if i != '-1'])\n",
    "    \n",
    "  obs = df.groupby('bin_name').agg(\n",
    "      bin_start=('bin_start', 'first'),\n",
    "      bin_end=('bin_end', 'first'),\n",
    "      bin=('bin', 'first'),\n",
    "      bin_index=('bin_index', 'first'),\n",
    "      chrom_bin=('chrom_bin', 'first'),\n",
    "      degree=('read_name', 'nunique'),\n",
    "      genes=('gene_name', gene_list),\n",
    "      n_genes=('gene_name', n_genes),\n",
    "  ).reset_index()\n",
    "\n",
    "\n",
    "  # Ensure proper sorting using var_names\n",
    "  obs = obs.set_index('bin_index')\n",
    "  obs = obs.reindex(obs_names)\n",
    "  obs = obs.reset_index()\n",
    "  obs = obs.set_index('bin_name')\n",
    "\n",
    "  return obs\n",
    "\n",
    "obs = create_obs_df(df, \"\")\n",
    "obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a7916d7-0b5e-41c9-a18a-18e032da1e9a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2927a722-2e84-4606-807d-1c4510e971e6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "012b9909-1bc3-4889-b905-291a492c41cb",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "'break' outside loop (668683560.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[10], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    break\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m 'break' outside loop\n"
     ]
    }
   ],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "683ceb57-9d3b-4aa8-bd3a-17ac0029713d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_name</th>\n",
       "      <th>read_start</th>\n",
       "      <th>read_end</th>\n",
       "      <th>length_on_read</th>\n",
       "      <th>chrom</th>\n",
       "      <th>ref_start</th>\n",
       "      <th>ref_end</th>\n",
       "      <th>mapping_quality</th>\n",
       "      <th>basename</th>\n",
       "      <th>local_position</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_overlap</th>\n",
       "      <th>is_pt_gene</th>\n",
       "      <th>bin_start</th>\n",
       "      <th>bin_end</th>\n",
       "      <th>bin</th>\n",
       "      <th>chrom_bin</th>\n",
       "      <th>bin_name</th>\n",
       "      <th>bin_overlap</th>\n",
       "      <th>read_index</th>\n",
       "      <th>bin_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7796001</th>\n",
       "      <td>3891ee6d-53d1-4ee0-ba2f-3d22291d4493</td>\n",
       "      <td>0</td>\n",
       "      <td>4916</td>\n",
       "      <td>4916</td>\n",
       "      <td>9</td>\n",
       "      <td>121048825</td>\n",
       "      <td>121057172</td>\n",
       "      <td>60</td>\n",
       "      <td>batch01</td>\n",
       "      <td>121052998</td>\n",
       "      <td>...</td>\n",
       "      <td>8347</td>\n",
       "      <td>True</td>\n",
       "      <td>121000000</td>\n",
       "      <td>122000000</td>\n",
       "      <td>1394</td>\n",
       "      <td>121</td>\n",
       "      <td>chr9:121</td>\n",
       "      <td>8347</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14148112</th>\n",
       "      <td>66953ddf-e76d-4cdf-aaf8-be028a2d7b04</td>\n",
       "      <td>6352</td>\n",
       "      <td>9540</td>\n",
       "      <td>3188</td>\n",
       "      <td>19</td>\n",
       "      <td>26583825</td>\n",
       "      <td>26592064</td>\n",
       "      <td>60</td>\n",
       "      <td>batch01</td>\n",
       "      <td>26587944</td>\n",
       "      <td>...</td>\n",
       "      <td>8239</td>\n",
       "      <td>True</td>\n",
       "      <td>26000000</td>\n",
       "      <td>27000000</td>\n",
       "      <td>2436</td>\n",
       "      <td>26</td>\n",
       "      <td>chr19:26</td>\n",
       "      <td>8239</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3788352</th>\n",
       "      <td>ad5b2240-893f-4ed0-a157-c2be66d8d754</td>\n",
       "      <td>0</td>\n",
       "      <td>1919</td>\n",
       "      <td>1919</td>\n",
       "      <td>4</td>\n",
       "      <td>127067225</td>\n",
       "      <td>127074760</td>\n",
       "      <td>60</td>\n",
       "      <td>batch04</td>\n",
       "      <td>127070992</td>\n",
       "      <td>...</td>\n",
       "      <td>7535</td>\n",
       "      <td>True</td>\n",
       "      <td>127000000</td>\n",
       "      <td>128000000</td>\n",
       "      <td>665</td>\n",
       "      <td>127</td>\n",
       "      <td>chr4:127</td>\n",
       "      <td>7535</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7462945</th>\n",
       "      <td>1cbc7b8c-2ad8-4365-a57c-037fc41af923</td>\n",
       "      <td>1305</td>\n",
       "      <td>6701</td>\n",
       "      <td>5396</td>\n",
       "      <td>8</td>\n",
       "      <td>21218739</td>\n",
       "      <td>21225694</td>\n",
       "      <td>1</td>\n",
       "      <td>batch04</td>\n",
       "      <td>21222216</td>\n",
       "      <td>...</td>\n",
       "      <td>-21218740</td>\n",
       "      <td>False</td>\n",
       "      <td>21000000</td>\n",
       "      <td>22000000</td>\n",
       "      <td>1163</td>\n",
       "      <td>21</td>\n",
       "      <td>chr8:21</td>\n",
       "      <td>6955</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7463448</th>\n",
       "      <td>9af4c1b9-a795-43a1-9669-55aff68e3b9d</td>\n",
       "      <td>426</td>\n",
       "      <td>5786</td>\n",
       "      <td>5360</td>\n",
       "      <td>8</td>\n",
       "      <td>21241017</td>\n",
       "      <td>21247152</td>\n",
       "      <td>60</td>\n",
       "      <td>batch03</td>\n",
       "      <td>21244084</td>\n",
       "      <td>...</td>\n",
       "      <td>-21241018</td>\n",
       "      <td>False</td>\n",
       "      <td>21000000</td>\n",
       "      <td>22000000</td>\n",
       "      <td>1163</td>\n",
       "      <td>21</td>\n",
       "      <td>chr8:21</td>\n",
       "      <td>6135</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     read_name  read_start  read_end  \\\n",
       "7796001   3891ee6d-53d1-4ee0-ba2f-3d22291d4493           0      4916   \n",
       "14148112  66953ddf-e76d-4cdf-aaf8-be028a2d7b04        6352      9540   \n",
       "3788352   ad5b2240-893f-4ed0-a157-c2be66d8d754           0      1919   \n",
       "7462945   1cbc7b8c-2ad8-4365-a57c-037fc41af923        1305      6701   \n",
       "7463448   9af4c1b9-a795-43a1-9669-55aff68e3b9d         426      5786   \n",
       "\n",
       "          length_on_read chrom  ref_start    ref_end  mapping_quality  \\\n",
       "7796001             4916     9  121048825  121057172               60   \n",
       "14148112            3188    19   26583825   26592064               60   \n",
       "3788352             1919     4  127067225  127074760               60   \n",
       "7462945             5396     8   21218739   21225694                1   \n",
       "7463448             5360     8   21241017   21247152               60   \n",
       "\n",
       "         basename  local_position  ...  gene_overlap  is_pt_gene  bin_start  \\\n",
       "7796001   batch01       121052998  ...          8347        True  121000000   \n",
       "14148112  batch01        26587944  ...          8239        True   26000000   \n",
       "3788352   batch04       127070992  ...          7535        True  127000000   \n",
       "7462945   batch04        21222216  ...     -21218740       False   21000000   \n",
       "7463448   batch03        21244084  ...     -21241018       False   21000000   \n",
       "\n",
       "            bin_end   bin chrom_bin  bin_name  bin_overlap  read_index  \\\n",
       "7796001   122000000  1394       121  chr9:121         8347           0   \n",
       "14148112   27000000  2436        26  chr19:26         8239           1   \n",
       "3788352   128000000   665       127  chr4:127         7535           2   \n",
       "7462945    22000000  1163        21   chr8:21         6955           3   \n",
       "7463448    22000000  1163        21   chr8:21         6135           4   \n",
       "\n",
       "          bin_index  \n",
       "7796001           0  \n",
       "14148112          1  \n",
       "3788352           2  \n",
       "7462945           3  \n",
       "7463448           3  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "78a80826-919b-45ec-8194-41b3be8a1588",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>read_name</th>\n",
       "      <th>bin_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ulk4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>3891ee6d-53d1-4ee0-ba2f-3d22291d4493</td>\n",
       "      <td>chr9:121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Smarca2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>66953ddf-e76d-4cdf-aaf8-be028a2d7b04</td>\n",
       "      <td>chr19:26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Dlgap3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ad5b2240-893f-4ed0-a157-c2be66d8d754</td>\n",
       "      <td>chr4:127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Gm56531</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>3f354c45-5e48-4f6d-8c7e-05369432b344</td>\n",
       "      <td>chr12:8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Twsg1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>d4626feb-16a2-4aac-8145-53e89b60bf7c</td>\n",
       "      <td>chr17:66</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name    gene_biotype                             read_name  bin_name\n",
       "0      Ulk4  protein_coding  3891ee6d-53d1-4ee0-ba2f-3d22291d4493  chr9:121\n",
       "1   Smarca2  protein_coding  66953ddf-e76d-4cdf-aaf8-be028a2d7b04  chr19:26\n",
       "2    Dlgap3  protein_coding  ad5b2240-893f-4ed0-a157-c2be66d8d754  chr4:127\n",
       "3   Gm56531          lncRNA  3f354c45-5e48-4f6d-8c7e-05369432b344   chr12:8\n",
       "4     Twsg1  protein_coding  d4626feb-16a2-4aac-8145-53e89b60bf7c  chr17:66"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_map = df[['gene_name', 'gene_biotype', 'read_name', 'bin_name']].drop_duplicates()\n",
    "gene_map = gene_map[gene_map['gene_name'] != '-1']\n",
    "gene_map = gene_map.reset_index(drop=True)\n",
    "\n",
    "gene_map.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7dd8777-8fe7-4b3f-be52-a11cd1df1a17",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1fbc7d4-6c81-4c4a-b780-4f059f5fcdbb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a894d735-2775-461f-a8f3-1fd21b972c83",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f345b00-5a87-4efa-9572-2de95f90f162",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "322045b4-0a86-4ccf-836a-e61dba2085eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['duplicate_maps'] = df.groupby(['read_name', 'read_start'])['gene_name'].transform('nunique')\n",
    "test = df[df['duplicate_maps'] > 1]\n",
    "test = test.sort_values(by=['read_name', 'ref_start'])\n",
    "print(f\"{test.shape=}\")\n",
    "test[['read_name', 'read_start', 'ref_start', 'chrom', 'gene_name', 'gene_biotype', 'gene_start']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c82465d-4755-48a4-8ee3-3c92a4ed7169",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f82a99c-83fe-4cb5-bcc0-3f5d78bd3440",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1b42c83-07ad-4bc3-b5d0-19417cd77bed",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['gene_name'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a44c732-9b49-414a-8d8b-96e127b42abb",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['read_name'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "402519d7-ac56-491d-8ca6-6d8bd13d4721",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01aae936-bdc3-4234-bd0e-49b6ed79f085",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fab9b082-f528-40b2-9bcb-2e8dd3e39242",
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = pd.read_parquet(gene_path)\n",
    "print(f\"{gdf.shape=}\")\n",
    "gdf.head()\n",
    "\n",
    "# gdf_pr = pr.PyRanges(gdf)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "482afa93-cbee-4b02-9971-d3a0469df79d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a765821-5e5c-4d6f-8af7-a94d1b9859f1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy",
   "language": "python",
   "name": "scanpy"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
