{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "97fc1b38-7cdb-4762-934d-6637d7d94c52",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "import time\n",
    "import gget\n",
    "import scipy\n",
    "import matplotlib.patches as patches\n",
    "from scipy.sparse import csr_matrix\n",
    "import anndata as an\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import math\n",
    "from scipy.stats import fisher_exact\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.cm as cm\n",
    "from pycirclize import Circos\n",
    "from scipy.interpolate import splprep, splev\n",
    "import networkx as nx\n",
    "import random\n",
    "from importlib import reload\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.preprocessing import MultiLabelBinarizer\n",
    "import warnings\n",
    "from itertools import combinations\n",
    "import ot\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "from matplotlib.colors import ListedColormap\n",
    "\n",
    "import surprise as sup\n",
    "\n",
    "from sklearn.decomposition import TruncatedSVD\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "\n",
    "\"\"\"WARNING: no warnings\"\"\"\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# local imports\n",
    "import anndata_utils as anntools\n",
    "\n",
    "source_path = os.path.abspath(\"../source/\")\n",
    "sys.path.append(source_path)\n",
    "import centrality as central\n",
    "import matrix\n",
    "import utils as ut\n",
    "import plotting as plt2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "314c3d24-fcf9-4179-8c0f-2641245a5305",
   "metadata": {},
   "source": [
    "# Load pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "894cac7f-f7c4-45fe-abd7-04401ce8baad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Gjb1',\n",
       " 'Amotl2',\n",
       " 'Yap1',\n",
       " 'Fbln1',\n",
       " 'Uaca',\n",
       " 'Antxr1',\n",
       " 'Fermt2',\n",
       " 'Serpinh1',\n",
       " 'Pls3',\n",
       " 'Gpx8']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def load_pathway(fpath):\n",
    "    \"\"\"\n",
    "    Loads an Enrichr-like database file into a boolean DataFrame.\n",
    "\n",
    "    Args:\n",
    "        fpath (str): Path to the Enrichr-like database file.\n",
    "\n",
    "    Returns:\n",
    "        pandas.DataFrame: A boolean DataFrame where:\n",
    "            - Index: Genes\n",
    "            - Columns: Pathways\n",
    "            - Values: True if the gene is in the pathway, False otherwise.\n",
    "    \"\"\"\n",
    "\n",
    "    result = []\n",
    "    with open(fpath,  encoding='utf-8') as f:\n",
    "        for line in f:\n",
    "            split_line = [x for x in line.strip().split('\\t') if x]  # Remove empty strings directly\n",
    "\n",
    "            row = {'label': split_line[0]}\n",
    "            for gene in split_line[1:]:\n",
    "                row[gene] = 1\n",
    "\n",
    "            result.append(row)\n",
    "\n",
    "    df = pd.DataFrame(result)\n",
    "    df = df.fillna(0.0).set_index('label').astype(bool).T  # Chained operations for clarity\n",
    "\n",
    "    return df\n",
    "\n",
    "fpath = \"../../ONT-single-cell/resources/PanglaoDB_Augmented_2021.txt\"\n",
    "pdf = load_pathway(fpath)\n",
    "stem_genes = list(pdf[pdf['Embryonic Stem Cells']].index)\n",
    "stem_genes = [x.title() for x in stem_genes]\n",
    "stem_genes[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fb2a2d52-813c-40c4-a2c2-72676ad2ab9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" LOAD the TF list \"\"\"\n",
    "fpath = \"/nfs/turbo/umms-indikar/shared/projects/twin_cell/data/b_matrix/SCENIC/scenic/tf_lists/allTFs_mm.txt\"\n",
    "tf_list = [x.strip() for x in open(fpath)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11f65ec9-a001-436d-8c60-40ca83b65c94",
   "metadata": {},
   "source": [
    "# Load expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a4a65750-228b-459f-9464-d445d6e71ddf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(raw) tdf.shape=(51883, 19)\n",
      "(filtered) tdf.shape=(51883, 19)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id(s)</th>\n",
       "      <th>length</th>\n",
       "      <th>effective_length</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>posterior_mean_count</th>\n",
       "      <th>posterior_standard_deviation_of_count</th>\n",
       "      <th>pme_TPM</th>\n",
       "      <th>pme_FPKM</th>\n",
       "      <th>TPM_ci_lower_bound</th>\n",
       "      <th>TPM_ci_upper_bound</th>\n",
       "      <th>TPM_coefficient_of_quartile_variation</th>\n",
       "      <th>FPKM_ci_lower_bound</th>\n",
       "      <th>FPKM_ci_upper_bound</th>\n",
       "      <th>FPKM_coefficient_of_quartile_variation</th>\n",
       "      <th>ens_gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>30929</th>\n",
       "      <td>ENSMUSG00000093416.1</td>\n",
       "      <td>ENSMUST00000176980.1</td>\n",
       "      <td>1748.00</td>\n",
       "      <td>1496.24</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.04</td>\n",
       "      <td>8.858160e-07</td>\n",
       "      <td>0.141585</td>\n",
       "      <td>0.657567</td>\n",
       "      <td>6.923890e-07</td>\n",
       "      <td>0.110727</td>\n",
       "      <td>0.657528</td>\n",
       "      <td>ENSMUSG00000093416</td>\n",
       "      <td>Gm18294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5906</th>\n",
       "      <td>ENSMUSG00000027075.16</td>\n",
       "      <td>ENSMUST00000028469.13,ENSMUST00000111624.7,ENS...</td>\n",
       "      <td>2298.15</td>\n",
       "      <td>2046.40</td>\n",
       "      <td>457.00</td>\n",
       "      <td>16.54</td>\n",
       "      <td>12.33</td>\n",
       "      <td>457.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>16.12</td>\n",
       "      <td>12.61</td>\n",
       "      <td>1.395610e+01</td>\n",
       "      <td>18.498400</td>\n",
       "      <td>0.048884</td>\n",
       "      <td>1.090210e+01</td>\n",
       "      <td>14.457400</td>\n",
       "      <td>0.048862</td>\n",
       "      <td>ENSMUSG00000027075</td>\n",
       "      <td>Slc43a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24952</th>\n",
       "      <td>ENSMUSG00000083394.2</td>\n",
       "      <td>ENSMUST00000118352.2</td>\n",
       "      <td>483.00</td>\n",
       "      <td>231.37</td>\n",
       "      <td>1.01</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.24</td>\n",
       "      <td>1.02</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.61</td>\n",
       "      <td>0.48</td>\n",
       "      <td>1.206700e-02</td>\n",
       "      <td>1.460530</td>\n",
       "      <td>0.476422</td>\n",
       "      <td>8.852180e-03</td>\n",
       "      <td>1.141890</td>\n",
       "      <td>0.476211</td>\n",
       "      <td>ENSMUSG00000083394</td>\n",
       "      <td>Gm11703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31109</th>\n",
       "      <td>ENSMUSG00000093711.1</td>\n",
       "      <td>ENSMUST00000176234.1</td>\n",
       "      <td>931.00</td>\n",
       "      <td>679.24</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.08</td>\n",
       "      <td>2.393220e-07</td>\n",
       "      <td>0.312362</td>\n",
       "      <td>0.653377</td>\n",
       "      <td>1.874730e-07</td>\n",
       "      <td>0.244212</td>\n",
       "      <td>0.653296</td>\n",
       "      <td>ENSMUSG00000093711</td>\n",
       "      <td>Vmn2r-ps125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12328</th>\n",
       "      <td>ENSMUSG00000041718.15</td>\n",
       "      <td>ENSMUST00000040338.8,ENSMUST00000070801.10,ENS...</td>\n",
       "      <td>2937.80</td>\n",
       "      <td>2686.05</td>\n",
       "      <td>4468.00</td>\n",
       "      <td>123.19</td>\n",
       "      <td>91.87</td>\n",
       "      <td>4468.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>119.40</td>\n",
       "      <td>93.35</td>\n",
       "      <td>1.118890e+02</td>\n",
       "      <td>126.778000</td>\n",
       "      <td>0.021625</td>\n",
       "      <td>8.760090e+01</td>\n",
       "      <td>99.226700</td>\n",
       "      <td>0.021644</td>\n",
       "      <td>ENSMUSG00000041718</td>\n",
       "      <td>Alg13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gene_id  \\\n",
       "30929   ENSMUSG00000093416.1   \n",
       "5906   ENSMUSG00000027075.16   \n",
       "24952   ENSMUSG00000083394.2   \n",
       "31109   ENSMUSG00000093711.1   \n",
       "12328  ENSMUSG00000041718.15   \n",
       "\n",
       "                                        transcript_id(s)   length  \\\n",
       "30929                               ENSMUST00000176980.1  1748.00   \n",
       "5906   ENSMUST00000028469.13,ENSMUST00000111624.7,ENS...  2298.15   \n",
       "24952                               ENSMUST00000118352.2   483.00   \n",
       "31109                               ENSMUST00000176234.1   931.00   \n",
       "12328  ENSMUST00000040338.8,ENSMUST00000070801.10,ENS...  2937.80   \n",
       "\n",
       "       effective_length  expected_count     TPM   FPKM  posterior_mean_count  \\\n",
       "30929           1496.24            0.00    0.00   0.00                  0.00   \n",
       "5906            2046.40          457.00   16.54  12.33                457.00   \n",
       "24952            231.37            1.01    0.32   0.24                  1.02   \n",
       "31109            679.24            0.00    0.00   0.00                  0.00   \n",
       "12328           2686.05         4468.00  123.19  91.87               4468.00   \n",
       "\n",
       "       posterior_standard_deviation_of_count  pme_TPM  pme_FPKM  \\\n",
       "30929                                   0.00     0.05      0.04   \n",
       "5906                                    0.00    16.12     12.61   \n",
       "24952                                   0.16     0.61      0.48   \n",
       "31109                                   0.00     0.10      0.08   \n",
       "12328                                   0.00   119.40     93.35   \n",
       "\n",
       "       TPM_ci_lower_bound  TPM_ci_upper_bound  \\\n",
       "30929        8.858160e-07            0.141585   \n",
       "5906         1.395610e+01           18.498400   \n",
       "24952        1.206700e-02            1.460530   \n",
       "31109        2.393220e-07            0.312362   \n",
       "12328        1.118890e+02          126.778000   \n",
       "\n",
       "       TPM_coefficient_of_quartile_variation  FPKM_ci_lower_bound  \\\n",
       "30929                               0.657567         6.923890e-07   \n",
       "5906                                0.048884         1.090210e+01   \n",
       "24952                               0.476422         8.852180e-03   \n",
       "31109                               0.653377         1.874730e-07   \n",
       "12328                               0.021625         8.760090e+01   \n",
       "\n",
       "       FPKM_ci_upper_bound  FPKM_coefficient_of_quartile_variation  \\\n",
       "30929             0.110727                                0.657528   \n",
       "5906             14.457400                                0.048862   \n",
       "24952             1.141890                                0.476211   \n",
       "31109             0.244212                                0.653296   \n",
       "12328            99.226700                                0.021644   \n",
       "\n",
       "              ens_gene_id    gene_name  \n",
       "30929  ENSMUSG00000093416      Gm18294  \n",
       "5906   ENSMUSG00000027075      Slc43a1  \n",
       "24952  ENSMUSG00000083394      Gm11703  \n",
       "31109  ENSMUSG00000093711  Vmn2r-ps125  \n",
       "12328  ENSMUSG00000041718        Alg13  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fpath = \"/scratch/indikar_root/indikar1/shared_data/higher_order/expression_table/rna_table.parquet\"\n",
    "tdf = pd.read_parquet(fpath)\n",
    "print(f\"(raw) {tdf.shape=}\")\n",
    "print(f\"(filtered) {tdf.shape=}\")\n",
    "\n",
    "expression_map = dict(zip(tdf['gene_name'].values, tdf['TPM'].values))\n",
    "\n",
    "tdf.sample(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb45049e-c62c-48b2-bd47-c6b3318e27ac",
   "metadata": {},
   "source": [
    "# Load core scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ca7fc2e7-ac4d-4f71-8f75-145b436f6308",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scores.shape=(2431, 36)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_name</th>\n",
       "      <th>bin_index</th>\n",
       "      <th>bin_start</th>\n",
       "      <th>bin_end</th>\n",
       "      <th>bin</th>\n",
       "      <th>chrom</th>\n",
       "      <th>chrom_bin</th>\n",
       "      <th>degree</th>\n",
       "      <th>genes</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>...</th>\n",
       "      <th>ce_pagerank</th>\n",
       "      <th>hge_singular_vector_1</th>\n",
       "      <th>hge_logexp_unweighted</th>\n",
       "      <th>hge_logexp_degree_weighted</th>\n",
       "      <th>hge_logexp_RNA_weighted</th>\n",
       "      <th>hge_logexp_ATAC_weighted</th>\n",
       "      <th>global_singular_vector_1</th>\n",
       "      <th>global_hge_logexp_unweighted</th>\n",
       "      <th>global_hge_logexp_RNA_weighted</th>\n",
       "      <th>is_core</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:3</td>\n",
       "      <td>2235</td>\n",
       "      <td>3000000</td>\n",
       "      <td>4000000</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2953</td>\n",
       "      <td>Gm37363;Gm37381;Gm7341;Gm38148;Gm37180;Gm19938...</td>\n",
       "      <td>15</td>\n",
       "      <td>...</td>\n",
       "      <td>0.117995</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.148099</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.174370</td>\n",
       "      <td>0.222380</td>\n",
       "      <td>0.330855</td>\n",
       "      <td>0.501567</td>\n",
       "      <td>0.167764</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:4</td>\n",
       "      <td>964</td>\n",
       "      <td>4000000</td>\n",
       "      <td>5000000</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>3331</td>\n",
       "      <td>Gm7369;A930006A01Rik;Gm37381;Gm38076;Mrpl15;Gm...</td>\n",
       "      <td>22</td>\n",
       "      <td>...</td>\n",
       "      <td>0.372869</td>\n",
       "      <td>0.103845</td>\n",
       "      <td>0.239187</td>\n",
       "      <td>0.281148</td>\n",
       "      <td>0.468270</td>\n",
       "      <td>0.560302</td>\n",
       "      <td>0.412235</td>\n",
       "      <td>0.609261</td>\n",
       "      <td>0.369136</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:5</td>\n",
       "      <td>1823</td>\n",
       "      <td>5000000</td>\n",
       "      <td>6000000</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>3341</td>\n",
       "      <td>Atp6v1h;Gm17101;Gm16041;Rgs20;Gm36965;Gm38264;...</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>0.224249</td>\n",
       "      <td>0.057865</td>\n",
       "      <td>0.647027</td>\n",
       "      <td>0.201096</td>\n",
       "      <td>0.243125</td>\n",
       "      <td>0.405148</td>\n",
       "      <td>0.425601</td>\n",
       "      <td>0.528269</td>\n",
       "      <td>0.216383</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1:6</td>\n",
       "      <td>1288</td>\n",
       "      <td>6000000</td>\n",
       "      <td>7000000</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>3422</td>\n",
       "      <td>St18;Gm7449;Gm19214;Gm37108;Gm2147;Gm19026;473...</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>0.534940</td>\n",
       "      <td>0.172229</td>\n",
       "      <td>0.760979</td>\n",
       "      <td>0.401754</td>\n",
       "      <td>0.250762</td>\n",
       "      <td>0.521777</td>\n",
       "      <td>0.448264</td>\n",
       "      <td>0.492633</td>\n",
       "      <td>0.258762</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1:7</td>\n",
       "      <td>144</td>\n",
       "      <td>7000000</td>\n",
       "      <td>8000000</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>3289</td>\n",
       "      <td>Gm23274;Gm5694;Rps2-ps2;Gm37225;Gm37489;Gm2690...</td>\n",
       "      <td>18</td>\n",
       "      <td>...</td>\n",
       "      <td>0.099600</td>\n",
       "      <td>0.053319</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.160235</td>\n",
       "      <td>0.190846</td>\n",
       "      <td>0.204818</td>\n",
       "      <td>0.414267</td>\n",
       "      <td>0.510526</td>\n",
       "      <td>0.204855</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  bin_name  bin_index  bin_start  bin_end  bin chrom  chrom_bin  degree  \\\n",
       "0   chr1:3       2235    3000000  4000000    3     1          3    2953   \n",
       "1   chr1:4        964    4000000  5000000    4     1          4    3331   \n",
       "2   chr1:5       1823    5000000  6000000    5     1          5    3341   \n",
       "3   chr1:6       1288    6000000  7000000    6     1          6    3422   \n",
       "4   chr1:7        144    7000000  8000000    7     1          7    3289   \n",
       "\n",
       "                                               genes  n_genes  ...  \\\n",
       "0  Gm37363;Gm37381;Gm7341;Gm38148;Gm37180;Gm19938...       15  ...   \n",
       "1  Gm7369;A930006A01Rik;Gm37381;Gm38076;Mrpl15;Gm...       22  ...   \n",
       "2  Atp6v1h;Gm17101;Gm16041;Rgs20;Gm36965;Gm38264;...       12  ...   \n",
       "3  St18;Gm7449;Gm19214;Gm37108;Gm2147;Gm19026;473...       10  ...   \n",
       "4  Gm23274;Gm5694;Rps2-ps2;Gm37225;Gm37489;Gm2690...       18  ...   \n",
       "\n",
       "   ce_pagerank  hge_singular_vector_1  hge_logexp_unweighted  \\\n",
       "0     0.117995               0.000000               0.148099   \n",
       "1     0.372869               0.103845               0.239187   \n",
       "2     0.224249               0.057865               0.647027   \n",
       "3     0.534940               0.172229               0.760979   \n",
       "4     0.099600               0.053319               0.000000   \n",
       "\n",
       "   hge_logexp_degree_weighted  hge_logexp_RNA_weighted  \\\n",
       "0                    0.000000                 0.174370   \n",
       "1                    0.281148                 0.468270   \n",
       "2                    0.201096                 0.243125   \n",
       "3                    0.401754                 0.250762   \n",
       "4                    0.160235                 0.190846   \n",
       "\n",
       "   hge_logexp_ATAC_weighted  global_singular_vector_1  \\\n",
       "0                  0.222380                  0.330855   \n",
       "1                  0.560302                  0.412235   \n",
       "2                  0.405148                  0.425601   \n",
       "3                  0.521777                  0.448264   \n",
       "4                  0.204818                  0.414267   \n",
       "\n",
       "   global_hge_logexp_unweighted  global_hge_logexp_RNA_weighted  is_core  \n",
       "0                      0.501567                        0.167764    False  \n",
       "1                      0.609261                        0.369136     True  \n",
       "2                      0.528269                        0.216383    False  \n",
       "3                      0.492633                        0.258762    False  \n",
       "4                      0.510526                        0.204855    False  \n",
       "\n",
       "[5 rows x 37 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resolution = 1000000\n",
    "score_column = 'global_hge_logexp_RNA_weighted'\n",
    "core_threshold_quantile = 0.75\n",
    "\n",
    "fpath = f\"/scratch/indikar_root/indikar1/shared_data/higher_order/global_core_score/population_mESC_{resolution}_scores.csv\"\n",
    "scores = pd.read_csv(fpath)\n",
    "print(f\"{scores.shape=}\")\n",
    "\n",
    "threshold = np.quantile(scores[score_column].fillna(0.0), core_threshold_quantile)\n",
    "scores['is_core'] = scores[score_column] > threshold\n",
    "\n",
    "scores.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "660a547e-ddde-44d6-a790-d4c5e3c4955b",
   "metadata": {},
   "source": [
    "# single-cell transcription factories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f11c513a-97ff-4ae9-92b9-7216af049784",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df.shape=(59016150, 13)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>read_name</th>\n",
       "      <th>bin_name</th>\n",
       "      <th>is_tf</th>\n",
       "      <th>is_marker</th>\n",
       "      <th>ATACSeq_1</th>\n",
       "      <th>CTCF</th>\n",
       "      <th>H3K27ac</th>\n",
       "      <th>H3K27me3</th>\n",
       "      <th>RNA_5</th>\n",
       "      <th>global_hge_logexp_RNA_weighted</th>\n",
       "      <th>expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Jazf1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>07e775f2-ee00-4bd0-a2e4-601ded549c44</td>\n",
       "      <td>chr6:52</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0.716487</td>\n",
       "      <td>0.916573</td>\n",
       "      <td>0.903254</td>\n",
       "      <td>2.147158</td>\n",
       "      <td>0.352027</td>\n",
       "      <td>0.222275</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Nmur1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>19a335b5-31d1-44a2-ae3e-73433ef3a519</td>\n",
       "      <td>chr1:86</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.803481</td>\n",
       "      <td>1.474600</td>\n",
       "      <td>2.044946</td>\n",
       "      <td>0.442640</td>\n",
       "      <td>4.862435</td>\n",
       "      <td>0.727237</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Nmur1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>e880790d-0a46-4fdf-b4d8-1c944a8c79bf</td>\n",
       "      <td>chr1:86</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.803481</td>\n",
       "      <td>1.474600</td>\n",
       "      <td>2.044946</td>\n",
       "      <td>0.442640</td>\n",
       "      <td>4.862435</td>\n",
       "      <td>0.727237</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Nmur1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>0293c95c-85d4-4db9-b18a-840971c78330</td>\n",
       "      <td>chr1:86</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.803481</td>\n",
       "      <td>1.474600</td>\n",
       "      <td>2.044946</td>\n",
       "      <td>0.442640</td>\n",
       "      <td>4.862435</td>\n",
       "      <td>0.727237</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Nmur1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>9610ddea-ac3e-4b49-a118-97485a0675d8</td>\n",
       "      <td>chr1:86</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.803481</td>\n",
       "      <td>1.474600</td>\n",
       "      <td>2.044946</td>\n",
       "      <td>0.442640</td>\n",
       "      <td>4.862435</td>\n",
       "      <td>0.727237</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name    gene_biotype                             read_name bin_name  \\\n",
       "0     Jazf1  protein_coding  07e775f2-ee00-4bd0-a2e4-601ded549c44  chr6:52   \n",
       "1     Nmur1  protein_coding  19a335b5-31d1-44a2-ae3e-73433ef3a519  chr1:86   \n",
       "2     Nmur1  protein_coding  e880790d-0a46-4fdf-b4d8-1c944a8c79bf  chr1:86   \n",
       "3     Nmur1  protein_coding  0293c95c-85d4-4db9-b18a-840971c78330  chr1:86   \n",
       "4     Nmur1  protein_coding  9610ddea-ac3e-4b49-a118-97485a0675d8  chr1:86   \n",
       "\n",
       "   is_tf  is_marker  ATACSeq_1      CTCF   H3K27ac  H3K27me3     RNA_5  \\\n",
       "0   True      False   0.716487  0.916573  0.903254  2.147158  0.352027   \n",
       "1  False      False   0.803481  1.474600  2.044946  0.442640  4.862435   \n",
       "2  False      False   0.803481  1.474600  2.044946  0.442640  4.862435   \n",
       "3  False      False   0.803481  1.474600  2.044946  0.442640  4.862435   \n",
       "4  False      False   0.803481  1.474600  2.044946  0.442640  4.862435   \n",
       "\n",
       "   global_hge_logexp_RNA_weighted  expression  \n",
       "0                        0.222275        0.08  \n",
       "1                        0.727237        0.08  \n",
       "2                        0.727237        0.08  \n",
       "3                        0.727237        0.08  \n",
       "4                        0.727237        0.08  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\" LOAD the genes \"\"\"\n",
    "fpath = f\"/scratch/indikar_root/indikar1/shared_data/higher_order/lightweight/singlecell_mESC_{resolution}_gdf.parquet\"\n",
    "df = pd.read_parquet(fpath)\n",
    "df['is_tf'] = df['gene_name'].isin(tf_list)\n",
    "df['is_marker'] = df['gene_name'].isin(stem_genes)\n",
    "\n",
    "\n",
    "\"\"\" MERGE the core scores \"\"\"\n",
    "merge_columns = [\n",
    "    'bin_name',\n",
    "    'ATACSeq_1',\n",
    "    'CTCF',\n",
    "    'H3K27ac', \n",
    "    'H3K27me3',\n",
    "    'RNA_5',\n",
    "    'global_hge_logexp_RNA_weighted'\n",
    "]\n",
    "\n",
    "df = pd.merge(\n",
    "    df, scores[merge_columns].fillna(0.0), how=\"left\",\n",
    ")\n",
    "\n",
    "\"\"\" MERGE expression information \"\"\"\n",
    "df['expression'] = df['gene_name'].map(expression_map)\n",
    "\n",
    "print(f\"{df.shape=}\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db5265cb-6505-404c-b41d-b72ef0b5b487",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8dc0a2f-6aa2-4703-b89e-029070930eff",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f64a68d-64b8-4ce4-a0b9-cd050a6fe597",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy",
   "language": "python",
   "name": "scanpy"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
