{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "97fc1b38-7cdb-4762-934d-6637d7d94c52",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "import time\n",
    "import gget\n",
    "import scipy\n",
    "import matplotlib.patches as patches\n",
    "from scipy.sparse import csr_matrix\n",
    "import anndata as an\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import math\n",
    "from scipy.stats import fisher_exact\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.cm as cm\n",
    "from pycirclize import Circos\n",
    "from scipy.interpolate import splprep, splev\n",
    "import networkx as nx\n",
    "import random\n",
    "from importlib import reload\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.preprocessing import MultiLabelBinarizer\n",
    "import warnings\n",
    "from itertools import combinations\n",
    "import ot\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "from matplotlib.colors import ListedColormap\n",
    "\n",
    "import surprise as sup\n",
    "\n",
    "from sklearn.decomposition import TruncatedSVD\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "\n",
    "\"\"\"WARNING: no warnings\"\"\"\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# local imports\n",
    "import anndata_utils as anntools\n",
    "\n",
    "source_path = os.path.abspath(\"../source/\")\n",
    "sys.path.append(source_path)\n",
    "import centrality as central\n",
    "import matrix\n",
    "import utils as ut\n",
    "import plotting as plt2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "314c3d24-fcf9-4179-8c0f-2641245a5305",
   "metadata": {},
   "source": [
    "# Load pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "894cac7f-f7c4-45fe-abd7-04401ce8baad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Gjb1',\n",
       " 'Amotl2',\n",
       " 'Yap1',\n",
       " 'Fbln1',\n",
       " 'Uaca',\n",
       " 'Antxr1',\n",
       " 'Fermt2',\n",
       " 'Serpinh1',\n",
       " 'Pls3',\n",
       " 'Gpx8']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def load_pathway(fpath):\n",
    "    \"\"\"\n",
    "    Loads an Enrichr-like database file into a boolean DataFrame.\n",
    "\n",
    "    Args:\n",
    "        fpath (str): Path to the Enrichr-like database file.\n",
    "\n",
    "    Returns:\n",
    "        pandas.DataFrame: A boolean DataFrame where:\n",
    "            - Index: Genes\n",
    "            - Columns: Pathways\n",
    "            - Values: True if the gene is in the pathway, False otherwise.\n",
    "    \"\"\"\n",
    "\n",
    "    result = []\n",
    "    with open(fpath,  encoding='utf-8') as f:\n",
    "        for line in f:\n",
    "            split_line = [x for x in line.strip().split('\\t') if x]  # Remove empty strings directly\n",
    "\n",
    "            row = {'label': split_line[0]}\n",
    "            for gene in split_line[1:]:\n",
    "                row[gene] = 1\n",
    "\n",
    "            result.append(row)\n",
    "\n",
    "    df = pd.DataFrame(result)\n",
    "    df = df.fillna(0.0).set_index('label').astype(bool).T  # Chained operations for clarity\n",
    "\n",
    "    return df\n",
    "\n",
    "fpath = \"../../ONT-single-cell/resources/PanglaoDB_Augmented_2021.txt\"\n",
    "pdf = load_pathway(fpath)\n",
    "stem_genes = list(pdf[pdf['Embryonic Stem Cells']].index)\n",
    "stem_genes = [x.title() for x in stem_genes]\n",
    "stem_genes[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fb2a2d52-813c-40c4-a2c2-72676ad2ab9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" LOAD the TF list \"\"\"\n",
    "fpath = \"/nfs/turbo/umms-indikar/shared/projects/twin_cell/data/b_matrix/SCENIC/scenic/tf_lists/allTFs_mm.txt\"\n",
    "tf_list = [x.strip() for x in open(fpath)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11f65ec9-a001-436d-8c60-40ca83b65c94",
   "metadata": {},
   "source": [
    "# Load expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a4a65750-228b-459f-9464-d445d6e71ddf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(raw) tdf.shape=(51883, 19)\n",
      "(filtered) tdf.shape=(51883, 19)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id(s)</th>\n",
       "      <th>length</th>\n",
       "      <th>effective_length</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>posterior_mean_count</th>\n",
       "      <th>posterior_standard_deviation_of_count</th>\n",
       "      <th>pme_TPM</th>\n",
       "      <th>pme_FPKM</th>\n",
       "      <th>TPM_ci_lower_bound</th>\n",
       "      <th>TPM_ci_upper_bound</th>\n",
       "      <th>TPM_coefficient_of_quartile_variation</th>\n",
       "      <th>FPKM_ci_lower_bound</th>\n",
       "      <th>FPKM_ci_upper_bound</th>\n",
       "      <th>FPKM_coefficient_of_quartile_variation</th>\n",
       "      <th>ens_gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6351</th>\n",
       "      <td>ENSMUSG00000027878.11</td>\n",
       "      <td>ENSMUST00000079812.7,ENSMUST00000198324.1</td>\n",
       "      <td>10500.0</td>\n",
       "      <td>10248.24</td>\n",
       "      <td>1047.0</td>\n",
       "      <td>7.57</td>\n",
       "      <td>5.64</td>\n",
       "      <td>1047.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.26</td>\n",
       "      <td>5.68</td>\n",
       "      <td>6.803730e+00</td>\n",
       "      <td>7.740130</td>\n",
       "      <td>0.021901</td>\n",
       "      <td>5.310430e+00</td>\n",
       "      <td>6.042870</td>\n",
       "      <td>0.021852</td>\n",
       "      <td>ENSMUSG00000027878</td>\n",
       "      <td>Notch2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51879</th>\n",
       "      <td>ENSMUSG00000118390.1</td>\n",
       "      <td>ENSMUST00000237599.1</td>\n",
       "      <td>2368.0</td>\n",
       "      <td>2116.24</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.03</td>\n",
       "      <td>2.025230e-07</td>\n",
       "      <td>0.099766</td>\n",
       "      <td>0.653571</td>\n",
       "      <td>1.580150e-07</td>\n",
       "      <td>0.078040</td>\n",
       "      <td>0.653555</td>\n",
       "      <td>ENSMUSG00000118390</td>\n",
       "      <td>Gm50102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39641</th>\n",
       "      <td>ENSMUSG00000105074.1</td>\n",
       "      <td>ENSMUST00000200605.1</td>\n",
       "      <td>466.0</td>\n",
       "      <td>214.41</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.33</td>\n",
       "      <td>0.26</td>\n",
       "      <td>1.976270e-06</td>\n",
       "      <td>0.988187</td>\n",
       "      <td>0.656864</td>\n",
       "      <td>1.542960e-06</td>\n",
       "      <td>0.772439</td>\n",
       "      <td>0.656955</td>\n",
       "      <td>ENSMUSG00000105074</td>\n",
       "      <td>Gm43818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18075</th>\n",
       "      <td>ENSMUSG00000064737.1</td>\n",
       "      <td>ENSMUST00000082803.1</td>\n",
       "      <td>111.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ENSMUSG00000064737</td>\n",
       "      <td>Gm23056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39366</th>\n",
       "      <td>ENSMUSG00000104759.2</td>\n",
       "      <td>ENSMUST00000197110.2</td>\n",
       "      <td>386.0</td>\n",
       "      <td>135.18</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.41</td>\n",
       "      <td>6.959990e-06</td>\n",
       "      <td>1.541170</td>\n",
       "      <td>0.651119</td>\n",
       "      <td>5.437820e-06</td>\n",
       "      <td>1.205300</td>\n",
       "      <td>0.651176</td>\n",
       "      <td>ENSMUSG00000104759</td>\n",
       "      <td>Ranbp2-ps12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gene_id                           transcript_id(s)  \\\n",
       "6351   ENSMUSG00000027878.11  ENSMUST00000079812.7,ENSMUST00000198324.1   \n",
       "51879   ENSMUSG00000118390.1                       ENSMUST00000237599.1   \n",
       "39641   ENSMUSG00000105074.1                       ENSMUST00000200605.1   \n",
       "18075   ENSMUSG00000064737.1                       ENSMUST00000082803.1   \n",
       "39366   ENSMUSG00000104759.2                       ENSMUST00000197110.2   \n",
       "\n",
       "        length  effective_length  expected_count   TPM  FPKM  \\\n",
       "6351   10500.0          10248.24          1047.0  7.57  5.64   \n",
       "51879   2368.0           2116.24             0.0  0.00  0.00   \n",
       "39641    466.0            214.41             0.0  0.00  0.00   \n",
       "18075    111.0              0.00             0.0  0.00  0.00   \n",
       "39366    386.0            135.18             0.0  0.00  0.00   \n",
       "\n",
       "       posterior_mean_count  posterior_standard_deviation_of_count  pme_TPM  \\\n",
       "6351                 1047.0                                    0.0     7.26   \n",
       "51879                   0.0                                    0.0     0.03   \n",
       "39641                   0.0                                    0.0     0.33   \n",
       "18075                   0.0                                    0.0     0.00   \n",
       "39366                   0.0                                    0.0     0.52   \n",
       "\n",
       "       pme_FPKM  TPM_ci_lower_bound  TPM_ci_upper_bound  \\\n",
       "6351       5.68        6.803730e+00            7.740130   \n",
       "51879      0.03        2.025230e-07            0.099766   \n",
       "39641      0.26        1.976270e-06            0.988187   \n",
       "18075      0.00        0.000000e+00            0.000000   \n",
       "39366      0.41        6.959990e-06            1.541170   \n",
       "\n",
       "       TPM_coefficient_of_quartile_variation  FPKM_ci_lower_bound  \\\n",
       "6351                                0.021901         5.310430e+00   \n",
       "51879                               0.653571         1.580150e-07   \n",
       "39641                               0.656864         1.542960e-06   \n",
       "18075                               0.000000         0.000000e+00   \n",
       "39366                               0.651119         5.437820e-06   \n",
       "\n",
       "       FPKM_ci_upper_bound  FPKM_coefficient_of_quartile_variation  \\\n",
       "6351              6.042870                                0.021852   \n",
       "51879             0.078040                                0.653555   \n",
       "39641             0.772439                                0.656955   \n",
       "18075             0.000000                                0.000000   \n",
       "39366             1.205300                                0.651176   \n",
       "\n",
       "              ens_gene_id    gene_name  \n",
       "6351   ENSMUSG00000027878       Notch2  \n",
       "51879  ENSMUSG00000118390      Gm50102  \n",
       "39641  ENSMUSG00000105074      Gm43818  \n",
       "18075  ENSMUSG00000064737      Gm23056  \n",
       "39366  ENSMUSG00000104759  Ranbp2-ps12  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fpath = \"/scratch/indikar_root/indikar1/shared_data/higher_order/expression_table/rna_table.parquet\"\n",
    "tdf = pd.read_parquet(fpath)\n",
    "print(f\"(raw) {tdf.shape=}\")\n",
    "print(f\"(filtered) {tdf.shape=}\")\n",
    "\n",
    "expression_map = dict(zip(tdf['gene_name'].values, tdf['TPM'].values))\n",
    "\n",
    "tdf.sample(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb45049e-c62c-48b2-bd47-c6b3318e27ac",
   "metadata": {},
   "source": [
    "# Load core scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ca7fc2e7-ac4d-4f71-8f75-145b436f6308",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scores.shape=(2431, 36)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_name</th>\n",
       "      <th>bin_index</th>\n",
       "      <th>bin_start</th>\n",
       "      <th>bin_end</th>\n",
       "      <th>bin</th>\n",
       "      <th>chrom</th>\n",
       "      <th>chrom_bin</th>\n",
       "      <th>degree</th>\n",
       "      <th>genes</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>...</th>\n",
       "      <th>ce_pagerank</th>\n",
       "      <th>hge_singular_vector_1</th>\n",
       "      <th>hge_logexp_unweighted</th>\n",
       "      <th>hge_logexp_degree_weighted</th>\n",
       "      <th>hge_logexp_RNA_weighted</th>\n",
       "      <th>hge_logexp_ATAC_weighted</th>\n",
       "      <th>global_singular_vector_1</th>\n",
       "      <th>global_hge_logexp_unweighted</th>\n",
       "      <th>global_hge_logexp_RNA_weighted</th>\n",
       "      <th>is_core</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:3</td>\n",
       "      <td>2235</td>\n",
       "      <td>3000000</td>\n",
       "      <td>4000000</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2953</td>\n",
       "      <td>Gm37363;Gm37381;Gm7341;Gm38148;Gm37180;Gm19938...</td>\n",
       "      <td>15</td>\n",
       "      <td>...</td>\n",
       "      <td>0.117995</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.148099</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.174370</td>\n",
       "      <td>0.222380</td>\n",
       "      <td>0.330855</td>\n",
       "      <td>0.501567</td>\n",
       "      <td>0.167764</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:4</td>\n",
       "      <td>964</td>\n",
       "      <td>4000000</td>\n",
       "      <td>5000000</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>3331</td>\n",
       "      <td>Gm7369;A930006A01Rik;Gm37381;Gm38076;Mrpl15;Gm...</td>\n",
       "      <td>22</td>\n",
       "      <td>...</td>\n",
       "      <td>0.372869</td>\n",
       "      <td>0.103845</td>\n",
       "      <td>0.239187</td>\n",
       "      <td>0.281148</td>\n",
       "      <td>0.468270</td>\n",
       "      <td>0.560302</td>\n",
       "      <td>0.412235</td>\n",
       "      <td>0.609261</td>\n",
       "      <td>0.369136</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:5</td>\n",
       "      <td>1823</td>\n",
       "      <td>5000000</td>\n",
       "      <td>6000000</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>3341</td>\n",
       "      <td>Atp6v1h;Gm17101;Gm16041;Rgs20;Gm36965;Gm38264;...</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>0.224249</td>\n",
       "      <td>0.057865</td>\n",
       "      <td>0.647027</td>\n",
       "      <td>0.201096</td>\n",
       "      <td>0.243125</td>\n",
       "      <td>0.405148</td>\n",
       "      <td>0.425601</td>\n",
       "      <td>0.528269</td>\n",
       "      <td>0.216383</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1:6</td>\n",
       "      <td>1288</td>\n",
       "      <td>6000000</td>\n",
       "      <td>7000000</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>3422</td>\n",
       "      <td>St18;Gm7449;Gm19214;Gm37108;Gm2147;Gm19026;473...</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>0.534940</td>\n",
       "      <td>0.172229</td>\n",
       "      <td>0.760979</td>\n",
       "      <td>0.401754</td>\n",
       "      <td>0.250762</td>\n",
       "      <td>0.521777</td>\n",
       "      <td>0.448264</td>\n",
       "      <td>0.492633</td>\n",
       "      <td>0.258762</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1:7</td>\n",
       "      <td>144</td>\n",
       "      <td>7000000</td>\n",
       "      <td>8000000</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>3289</td>\n",
       "      <td>Gm23274;Gm5694;Rps2-ps2;Gm37225;Gm37489;Gm2690...</td>\n",
       "      <td>18</td>\n",
       "      <td>...</td>\n",
       "      <td>0.099600</td>\n",
       "      <td>0.053319</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.160235</td>\n",
       "      <td>0.190846</td>\n",
       "      <td>0.204818</td>\n",
       "      <td>0.414267</td>\n",
       "      <td>0.510526</td>\n",
       "      <td>0.204855</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  bin_name  bin_index  bin_start  bin_end  bin chrom  chrom_bin  degree  \\\n",
       "0   chr1:3       2235    3000000  4000000    3     1          3    2953   \n",
       "1   chr1:4        964    4000000  5000000    4     1          4    3331   \n",
       "2   chr1:5       1823    5000000  6000000    5     1          5    3341   \n",
       "3   chr1:6       1288    6000000  7000000    6     1          6    3422   \n",
       "4   chr1:7        144    7000000  8000000    7     1          7    3289   \n",
       "\n",
       "                                               genes  n_genes  ...  \\\n",
       "0  Gm37363;Gm37381;Gm7341;Gm38148;Gm37180;Gm19938...       15  ...   \n",
       "1  Gm7369;A930006A01Rik;Gm37381;Gm38076;Mrpl15;Gm...       22  ...   \n",
       "2  Atp6v1h;Gm17101;Gm16041;Rgs20;Gm36965;Gm38264;...       12  ...   \n",
       "3  St18;Gm7449;Gm19214;Gm37108;Gm2147;Gm19026;473...       10  ...   \n",
       "4  Gm23274;Gm5694;Rps2-ps2;Gm37225;Gm37489;Gm2690...       18  ...   \n",
       "\n",
       "   ce_pagerank  hge_singular_vector_1  hge_logexp_unweighted  \\\n",
       "0     0.117995               0.000000               0.148099   \n",
       "1     0.372869               0.103845               0.239187   \n",
       "2     0.224249               0.057865               0.647027   \n",
       "3     0.534940               0.172229               0.760979   \n",
       "4     0.099600               0.053319               0.000000   \n",
       "\n",
       "   hge_logexp_degree_weighted  hge_logexp_RNA_weighted  \\\n",
       "0                    0.000000                 0.174370   \n",
       "1                    0.281148                 0.468270   \n",
       "2                    0.201096                 0.243125   \n",
       "3                    0.401754                 0.250762   \n",
       "4                    0.160235                 0.190846   \n",
       "\n",
       "   hge_logexp_ATAC_weighted  global_singular_vector_1  \\\n",
       "0                  0.222380                  0.330855   \n",
       "1                  0.560302                  0.412235   \n",
       "2                  0.405148                  0.425601   \n",
       "3                  0.521777                  0.448264   \n",
       "4                  0.204818                  0.414267   \n",
       "\n",
       "   global_hge_logexp_unweighted  global_hge_logexp_RNA_weighted  is_core  \n",
       "0                      0.501567                        0.167764    False  \n",
       "1                      0.609261                        0.369136     True  \n",
       "2                      0.528269                        0.216383    False  \n",
       "3                      0.492633                        0.258762    False  \n",
       "4                      0.510526                        0.204855    False  \n",
       "\n",
       "[5 rows x 37 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resolution = 1000000\n",
    "score_column = 'global_hge_logexp_RNA_weighted'\n",
    "core_threshold_quantile = 0.75\n",
    "\n",
    "fpath = f\"/scratch/indikar_root/indikar1/shared_data/higher_order/global_core_score/population_mESC_{resolution}_scores.csv\"\n",
    "scores = pd.read_csv(fpath)\n",
    "print(f\"{scores.shape=}\")\n",
    "\n",
    "threshold = np.quantile(scores[score_column].fillna(0.0), core_threshold_quantile)\n",
    "scores['is_core'] = scores[score_column] > threshold\n",
    "\n",
    "scores.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "522e8b57-2730-45a5-99f3-2444b07d8b2d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dupes.shape=(10902736, 8)\n",
      "(raw) cells.shape=(339359, 8)\n",
      "(filtered) cells.shape=(339359, 2)\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'unique'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/.local/lib/python3.12/site-packages/pandas/core/indexes/base.py:3653\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3652\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3653\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3654\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32m~/.local/lib/python3.12/site-packages/pandas/_libs/index.pyx:147\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/.local/lib/python3.12/site-packages/pandas/_libs/index.pyx:176\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7080\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7088\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'unique'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[6], line 10\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m(filtered) \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mcells\u001b[38;5;241m.\u001b[39mshape\u001b[38;5;132;01m=}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      9\u001b[0m cell_map \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mdict\u001b[39m(\u001b[38;5;28mzip\u001b[39m(cells[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mread_id\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mvalues, cells[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcell_id\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mvalues))\n\u001b[0;32m---> 10\u001b[0m dup_map \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mdict\u001b[39m(\u001b[38;5;28mzip\u001b[39m(cells[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mread_id\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mvalues, \u001b[43mcells\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43munique\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mvalues))\n\u001b[1;32m     12\u001b[0m dupes\u001b[38;5;241m.\u001b[39mhead()\n",
      "File \u001b[0;32m~/.local/lib/python3.12/site-packages/pandas/core/frame.py:3761\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3759\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m   3760\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> 3761\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3762\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[1;32m   3763\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n",
      "File \u001b[0;32m~/.local/lib/python3.12/site-packages/pandas/core/indexes/base.py:3655\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3653\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine\u001b[38;5;241m.\u001b[39mget_loc(casted_key)\n\u001b[1;32m   3654\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[0;32m-> 3655\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3656\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3657\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3658\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3659\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3660\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'unique'"
     ]
    }
   ],
   "source": [
    "fpath = f\"/scratch/indikar_root/indikar1/shared_data/higher_order/duplicate_table/singlecell_mESC_{resolution}_dupes.parquet\"\n",
    "dupes = pd.read_parquet(fpath)\n",
    "print(f\"{dupes.shape=}\")\n",
    "\n",
    "cells = dupes[dupes['unique']]\n",
    "print(f\"(raw) {cells.shape=}\")\n",
    "cells = cells[['read_id', 'cell_id']].drop_duplicates()\n",
    "print(f\"(filtered) {cells.shape=}\")\n",
    "cell_map = dict(zip(cells['read_id'].values, cells['cell_id'].values))\n",
    "\n",
    "dupes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e437ecee-5b8b-4c33-b424-714ed1e2d2a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['read_id', 'mapping_quality', 'cell_id', 'exactly_unique',\n",
       "       'approximately_unique', 'read_group', 'unique', 'chrom'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dupes.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "660a547e-ddde-44d6-a790-d4c5e3c4955b",
   "metadata": {},
   "source": [
    "# single-cell transcription factories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f11c513a-97ff-4ae9-92b9-7216af049784",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" LOAD the genes \"\"\"\n",
    "fpath = f\"/scratch/indikar_root/indikar1/shared_data/higher_order/lightweight/singlecell_mESC_{resolution}_gdf.parquet\"\n",
    "df = pd.read_parquet(fpath)\n",
    "df['is_tf'] = df['gene_name'].isin(tf_list)\n",
    "df['is_marker'] = df['gene_name'].isin(stem_genes)\n",
    "\n",
    "\n",
    "\"\"\" MERGE the core scores \"\"\"\n",
    "merge_columns = [\n",
    "    'bin_name',\n",
    "    'ATACSeq_1',\n",
    "    'CTCF',\n",
    "    'H3K27ac', \n",
    "    'H3K27me3',\n",
    "    'RNA_5',\n",
    "    'global_hge_logexp_RNA_weighted'\n",
    "]\n",
    "\n",
    "df = pd.merge(\n",
    "    df, scores[merge_columns].fillna(0.0), how=\"left\",\n",
    ")\n",
    "\n",
    "\"\"\" MERGE expression information and cell_ids \"\"\"\n",
    "df['expression'] = df['gene_name'].map(expression_map)\n",
    "df['cell_id'] = df['read_name'].map(cell_map)\n",
    "df['unique'] = df['read_name'].map(dup_map)\n",
    "\n",
    "\n",
    "print(f\"{df.shape=}\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9c5c55d-8205-4be6-8fed-09a06712817a",
   "metadata": {},
   "source": [
    "# define single-cell transcription clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dbc056f-4a7b-4b2c-9ec1-0715886df4dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "order_threshold = 3\n",
    "expression_threshold = 0.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d08b107c-55fb-40aa-863d-f675974dfde1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "259e8078-f6d5-45bc-8e8c-9c0d666afa60",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be7e455f-8a0f-42a3-a9a2-c756ce5d3dd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db5265cb-6505-404c-b41d-b72ef0b5b487",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get hyperedges\n",
    "def get_hyperedge(genes):\n",
    "    return \"-\".join(sorted(genes))\n",
    "\n",
    "gx = df.groupby(['read_name', 'basename']).agg(\n",
    "    genes=('gene_name', get_hyperedge)\n",
    ").reset_index(drop=False)\n",
    "\n",
    "print(f\"{gx.shape=}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8dc0a2f-6aa2-4703-b89e-029070930eff",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f64a68d-64b8-4ce4-a0b9-cd050a6fe597",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy",
   "language": "python",
   "name": "scanpy"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
